### PR TITLE
nanocoap_fs: add nanoCoAP as VFS backend (remote CoAP fs)

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

With `aiocoap-fileserver` / `nanocoap_fileserver` we can export a local fs via CoAP.
This adds a module that does the opposite: Locally mount a remote filesystem that was exported via CoAP.

### Testing procedure

 - run `aiocoap-fileserver` in a directory of your choice
 - run `tests/net/nanocoap_fs`:

```
2024-05-22 20:33:23,394 # mount coap://[fe80::607f:b1ff:fef7:689c]/ /coap

2024-05-22 20:33:24,890 # ls /coap
2024-05-22 20:33:24,911 # .cargo/
2024-05-22 20:33:24,918 # Makefile.features
2024-05-22 20:33:24,925 # uncrustify-riot.cfg
2024-05-22 20:33:24,932 # kconfigs/
2024-05-22 20:33:24,937 # Makefile.base
2024-05-22 20:33:24,945 # .git/
2024-05-22 20:33:24,952 # tests/
2024-05-22 20:33:24,955 # CODEOWNERS
2024-05-22 20:33:24,959 # .gitattributes
2024-05-22 20:33:24,963 # .mailmap
2024-05-22 20:33:24,973 # sys/
2024-05-22 20:33:24,987 # examples/
2024-05-22 20:33:24,992 # tmp/
2024-05-22 20:33:25,001 # MAINTAINING.md
2024-05-22 20:33:25,011 # README.md
2024-05-22 20:33:25,020 # fuzzing/
2024-05-22 20:33:25,025 # features.yaml
2024-05-22 20:33:25,035 # dist/
2024-05-22 20:33:25,042 # Kconfig
2024-05-22 20:33:25,055 # SUBSYSTEMS.md
2024-05-22 20:33:25,066 # CONTRIBUTING.md
2024-05-22 20:33:25,082 # .murdock
2024-05-22 20:33:25,087 # doc.txt
2024-05-22 20:33:25,097 # .bandit
2024-05-22 20:33:25,108 # release-notes.txt
2024-05-22 20:33:25,118 # Vagrantfile
2024-05-22 20:33:25,128 # SECURITY.md
2024-05-22 20:33:25,139 # Makefile.dep
2024-05-22 20:33:25,151 # core/
2024-05-22 20:33:25,165 # .vscode/
2024-05-22 20:33:25,179 # drivers/
2024-05-22 20:33:25,190 # LOSTANDFOUND.md
2024-05-22 20:33:25,203 # .github/
2024-05-22 20:33:25,218 # build/
2024-05-22 20:33:25,234 # makefiles/
2024-05-22 20:33:25,249 # .gitignore
2024-05-22 20:33:25,263 # Makefile.include
2024-05-22 20:33:25,297 # pkg/
2024-05-22 20:33:25,313 # CODE_OF_CONDUCT.md
2024-05-22 20:33:25,326 # Makefile
2024-05-22 20:33:25,340 # CODING_CONVENTIONS_C++.md
2024-05-22 20:33:25,360 # bootloaders/
2024-05-22 20:33:25,380 # boards/
2024-05-22 20:33:25,395 # LICENSE
2024-05-22 20:33:25,409 # doc/
2024-05-22 20:33:25,428 # cpu/
2024-05-22 20:33:25,445 # CODING_CONVENTIONS.md
2024-05-22 20:33:25,461 # bors.toml
2024-05-22 20:33:25,482 # .murdock.yml
2024-05-22 20:33:25,500 # CITATION.cff
2024-05-22 20:33:25,513 # total 30 files

2024-05-22 20:35:10,538 # vfs r /coap/README.md 64
2024-05-22 20:35:10,544 # 00000000: 5b21 5b4e 6967 6874 6c79 2043 4920 7374  [![Nightly CI st
2024-05-22 20:35:10,548 # 00000010: 6174 7573 206d 6173 7465 725d 5b6d 6173  atus master][mas
2024-05-22 20:35:10,551 # 00000020: 7465 722d 6369 2d62 6164 6765 5d5d 5b6d  ter-ci-badge]][m
2024-05-22 20:35:10,558 # 00000030: 6173 7465 722d 6369 2d6c 696e 6b5d 0a5b  aster-ci-link].[
```

### Issues/PRs references

requires #20688 when `nanocoap_fileserver` is used as the server.

### CODE REVIEW

File: sys/include/vfs.h [hunk 1]
```diff
@@ -82,11 +85,11 @@ extern "C" {
 #define _MAX(a, b) ((a) > (b) ? (a) : (b))
 #endif
 
-#ifndef MAX5
+#ifndef MAX6
 /**
  * @brief   MAX5 Function to get the largest of 5 values
  */
-#define MAX5(a, b, c, d, e) _MAX(_MAX(_MAX((a), (b)), _MAX((c),(d))), (e))
+#define MAX6(a, b, c, d, e, f) _MAX(_MAX(_MAX(_MAX((a), (b)), _MAX((c),(d))), (e)), (f))
```
Comment (maribu) [inline sys/include/vfs.h:92]:
```suggestion
#define MAX6(a, b, c, d, e, f) MAX(MAX(MAX(MAX((a), (b)), MAX((c), (d))), (e)), (f))
```

Comment (maribu) [inline sys/include/vfs.h:92]:
```suggestion
#define MAX6(a, b, c, d, e, f) MAX(MAX(MAX(MAX((a), (b)), MAX((c), (d))), (e)), (f))
```

There's a `MAX()` in `#include "macros/utils.h"` these days that can be used. Also: The comment above needs an update.

File: sys/net/application_layer/nanocoap/fs.c [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+            buf->st_mode = S_IFDIR;
+        }
+    }
+
+out:
+    mutex_unlock(&fs->lock);
+    return res;
+}
+
+
```
Comment (maribu) [inline sys/net/application_layer/nanocoap/fs.c:222]:
```suggestion

```

File: sys/include/net/nanocoap/fs.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * @brief nanoCoAP file system configuration
+ */
+typedef struct {
+    const char *url;
+    nanocoap_sock_t sock;
+    mutex_t lock;
+    char urlbuf[CONFIG_SOCK_URLPATH_MAXLEN];
+} nanocoap_fs_t;
```
Comment (maribu) [inline sys/include/net/nanocoap/fs.h:38]: The CI insists on the struct members being documented.

File: sys/include/net/nanocoap/fs.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+typedef struct {
+    char urlbuf[CONFIG_SOCK_URLPATH_MAXLEN];
+    size_t offset;
+} nanocoap_fs_file_t;
```
Comment (maribu) [inline sys/include/net/nanocoap/fs.h:43]: Again: CI insists on more doc

File: sys/include/net/nanocoap/fs.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+typedef struct {
+    char urlbuf[CONFIG_SOCK_URLPATH_MAXLEN];
+    size_t offset;
+} nanocoap_fs_dir_t;
```
Comment (maribu) [inline sys/include/net/nanocoap/fs.h:48]: Same again :)

**Review by maribu — COMMENTED** [review]
This PR is surprisingly small for the feature it provides :)

File: sys/include/vfs.h [hunk 6]
```diff
@@ -208,6 +203,19 @@ extern "C" {
 #endif
 /** @} */
 
+/**
+ * @brief   VFS parameters for nanoCoAP FS
+ * @{
+ */
+#if defined(MODULE_NANOCOAP_FS) || DOXYGEN
+#  define NANOCOAP_FS_VFS_DIR_BUFFER_SIZE   (4 + CONFIG_SOCK_URLPATH_MAXLEN) /**< sizeof(nanocoap_fs_dir_t) */
```
Comment (maribu) [inline sys/include/vfs.h:211]:
```suggestion
#  define NANOCOAP_FS_VFS_DIR_BUFFER_SIZE \
          (4 + CONFIG_SOCK_URLPATH_MAXLEN) /**< sizeof(nanocoap_fs_dir_t) */
```

File: sys/include/vfs.h [hunk 7]
```diff
@@ -208,6 +203,19 @@ extern "C" {
...
+#  define NANOCOAP_FS_VFS_FILE_BUFFER_SIZE  (4 + CONFIG_SOCK_URLPATH_MAXLEN) /**< sizeof(nanocoap_fs_file_t) */
```
Comment (maribu) [inline sys/include/vfs.h:212]:
```suggestion
#  define NANOCOAP_FS_VFS_FILE_BUFFER_SIZE \
          (4 + CONFIG_SOCK_URLPATH_MAXLEN) /**< sizeof(nanocoap_fs_file_t) */
```

**Review by maribu — APPROVED** [review]
lgtm. Maybe make the CI happy regarding the line length limit?

**Comment by benpicco** [comment]
Thank you for the quick review!

**Pull request merged**